### PR TITLE
Fix PermissionError swallowing in _atomic_save

### LINKED ISSUE #21800: Checkpoints silently fail to save due to swallowed PermissionError in `_atomic_save`
### Bug description

`_atomic_save` was silently swallowing non-cross-device PermissionErrors, causing saves to report success while failing to write any checkpoints. Because training continues as if the process were successful, this is a data-loss risk.

### What version are you seeing the problem on?

v2.6

### Reproduced in studio

_No response_

### How to reproduce the bug

```python
import os
from pathlib import Path
import sys
import tempfile
import torch
from lightning.fabric.utilities.cloud_io import _atomic_save


def main():
    with tempfile.TemporaryDirectory() as tmpdir:
        # Create a read-only directory (mode 0o555: read & execute only, no write permission)
        readonly_dir = Path(tmpdir) / "readonly"
        readonly_dir.mkdir()
        os.chmod(readonly_dir, 0o555)

        filepath = readonly_dir / "test.ckpt"
        print(f"Attempting to save checkpoint to read-only path: {filepath}")

        try:
            # _atomic_save attempts to save inside the read-only directory.
            # A PermissionError will occur inside fsspec, but without the fix,
            # _atomic_save silently swallows it and returns None.
            _atomic_save({"key": torch.tensor([1, 2, 3])}, filepath)

            print("\n❌ BUG REPRODUCED: `_atomic_save` returned silently without raising an error!")
            print(f"   Did the file actually get saved? {filepath.exists()}")
            # Exit with code 1 to indicate the bug is present
            sys.exit(1)

        except PermissionError as e:
            print(f"\n✅ EXPECTED BEHAVIOR: PermissionError was properly raised:\n   {e}")
            sys.exit(0)
        finally:
            # Restore permissions so temporary directory cleanup succeeds
            os.chmod(readonly_dir, 0o777)


if __name__ == "__main__":
    main()
```

### Error messages and logs

```
# Error messages and logs here please
```

### Environment

<details>
  <summary>Current environment</summary>

```
#- PyTorch Lightning Version (e.g., 2.6.0):
#- PyTorch Version (e.g., 2.5):
#- Python version (e.g., 3.12):
#- OS (e.g., Linux):
#- CUDA/cuDNN version:
#- GPU models and configuration:
#- How you installed Lightning(`conda`, `pip`, source):
```

</details>

### More info

_No response_

cc @ethanwharris @lantiga

### PR DESCRIPTION
**Pull request by zhixiangli**
## What does this PR do?

Fixes https://github.com/Lightning-AI/pytorch-lightning/issues/21800

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

### CODE REVIEW

**Comment by zhixiangli** [comment]
Hey @deependujha, just bumping this when you get a moment to review.

**Review by deependujha — APPROVED** [review]

**Comment by zhixiangli** [comment]
@deependujha could we please go ahead and merge this pull request now that it has your approval? Thank you so much!

**Review by deependujha — APPROVED** [review]

**Comment by zhixiangli** [comment]
@tchaton Whenever you get a chance, could you give this a quick merge if everything looks good to you?

**Review by bhimrazy — APPROVED** [review]
LGTM, Thanks @zhixiangli

**Comment by zhixiangli** [comment]
Hi @bhimrazy and @deependujha! Gently bumping this PR as it looks like it’s approved and ready to go. Whenever you have a quick moment to merge, that would be greatly appreciated. Thanks!

**Review by ethanwharris — APPROVED** [review]
